test/java/integration/AndroidEngineSocketTagTest.java

```diff
@@ -0,0 +1,188 @@
+import io.envoyproxy.envoymobile.UpstreamHttpProtocol;
+import io.envoyproxy.envoymobile.engine.AndroidJniLibrary;
+import io.envoyproxy.envoymobile.engine.testing.RequestScenario;
+import io.envoyproxy.envoymobile.engine.testing.Response;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import okhttp3.mockwebserver.Dispatcher;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
+public class AndroidEngineSocketTagTest {
+
+  private final MockWebServer mockWebServer = new MockWebServer();
+  private Engine engine;
+
+  @BeforeClass
+  public static void loadJniLibrary() {
+    AndroidJniLibrary.loadTestLibrary();
+  }
+
+  @Before
+  public void setUpEngine() throws Exception {
+    CountDownLatch latch = new CountDownLatch(1);
+    Context appContext = ApplicationProvider.getApplicationContext();
+    engine = new AndroidEngineBuilder(appContext)
+                 .addLogLevel(LogLevel.OFF)
+                 .enableSocketTagging(true)
+                 .setOnEngineRunning(() -> {
+                   latch.countDown();
+                   return null;
+                 })
+                 .build();
+    latch.await(); // Don't launch a request before initialization has completed.
+  }
+
+  @After
+  public void shutdownEngine() throws Exception {
+    engine.terminate();
+    mockWebServer.shutdown();
+  }
+
+  @Test
+  public void socket_tag() throws Exception {
+    mockWebServer.setDispatcher(new Dispatcher() {
+      @Override
+      public MockResponse dispatch(RecordedRequest recordedRequest) {
+        assertThat(recordedRequest.getMethod()).isEqualTo(RequestMethod.GET.name());
+        assertThat(recordedRequest.getHeader("x-envoy-mobile-socket-tag")).isEqualTo(null);
+        return new MockResponse().setBody("This is my response Body");
+      }
+    });
+    mockWebServer.start();
+    RequestScenario requestScenario = new RequestScenario()
+                                          .setHttpMethod(RequestMethod.GET)
+                                          .setUrl(mockWebServer.url("post/flowers").toString())
+                                          .addHeader("x-envoy-mobile-socket-tag", "0,0");
```

Comment:
`RequestScenario` becomes a little bit of a leaky abstraction as in real application we are not supposed to call `addHeader("x-envoy-mobile-socket-tag", "0,0");` directly (since we added addSocketTag). 

In fact, I am not sure whether this test works the way it's intended too as it tries to add `x-envoy-mobile-socket-tag` header in https://github.com/envoyproxy/envoy-mobile/blob/0d7b43e30024980cb9cd438aecf93c56c3707e46/test/java/io/envoyproxy/envoymobile/engine/testing/RequestScenario.java#L53 and `add(...)` call does not accept headers that start with `x-envoy-mobile` prefix https://github.com/envoyproxy/envoy-mobile/blob/0d7b43e30024980cb9cd438aecf93c56c3707e46/library/kotlin/io/envoyproxy/envoymobile/HeadersBuilder.kt#L80

Maybe we could

Reply:
Blergh! Yes, good point. Ok, I've added an addSocketTag() method to RequestScenario which delegates to the underlying builder. I verified (via printf) that the socket tagging code is in fact called.

Reply:
great, thank you - I think that adding `addSocketTag` method to the test scenario is a way to go 👍 . That being said, I think that the right final solution in here is to change the underlying implementation of `TestScenario` so that it stores test state inside of `RequestHeadersBuilder` as opposed to on-the-side headers map. Only with `RequestHeadersBuilder` we can ensure that our tests simulate real-life condition.

I do not think that this has to be done as part of your PR so I am fine with it being merged the way it is now.

Reply:
>Or is it not the socket tagging stuff that is problematic, but rather the general header stuff?

Yes, that's what I meant

>Are you thinking we should get rid of the headers member in RequestScenario and use the RequestHeadersBuilder directly?

That was my idea. This way our tests would be guaranteed to "get an update" every time we update `RequestHeadersBuilder`.